doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -848,37 +1250,436 @@ The Ltac2 parser can be extended with syntactic notations.
+    items must be separated by :token:`string`.  For zero items, the associated value
+    is an empty list.
+
+  :n:`seq({+, @ltac2_scope })`
+    Parses the :token:`ltac2_scope`\s in order.  The associated value is a tuple,
+    omitting :token:`ltac2_scope`\s that are :token:`string`\s.
+    `self` and `next` are not permitted within `seq`.
+
+The following classes represent nonterminals with some special handling.  The
+table further down lists the classes that that are handled plainly.
+
+  :n:`constr {? ( {+, @scope_key } ) }`
+    Parses a :token:`term`.  If specified, the :token:`scope_key`\s are used to interpret
+    the term (as described in  :ref:`LocalInterpretationRulesForNotations`).  The last
+    :token:`scope_key` is the top of the scope stack that's applied to the :token:`term`.
+
+  :n:`open_constr`
+    Parses an open :token:`term`.
+
+  :n:`ident`
+    Parses :token:`ident` or :n:`$@ident`.  The first form returns :n:`ident:(@ident)`,
+    while the latter form returns the variable :n:`@ident`.
+
+  :n:`@string`
+    Accepts the specified string that is not a keyword, returning a value of `()`.
+
+  :n:`keyword(@string)`
+    Accepts the specified string that is a keyword, returning a value of `()`.
+
+  :n:`terminal(@string)`
+    Accepts the specified string whether it's a keyword or not, returning a value of `()`.
+
+  :n:`tactic {? (@int) }`
+    Parses an :token:`ltac2_expr`.  If :token:`int` is specified, the construct
+    parses a :n:`ltac2_expr@int`, for example `tactic(5)` parses :token:`ltac2_expr5`.
+    `tactic(6)` parses :token:`ltac2_expr`.
+    :token:`int` must be in the range `0 .. 6`.
+
+    You can also use `tactic` to accept an :token:`int` or a :token:`string`, but there's
+    no syntactic class that accepts *only* an :token:`int` or a :token:`string`.
+
+    .. todo this doesn't work as expected: "::" is in ltac2_expr1
+       Ltac2 Notation "ex4" x(tactic(0)) := x.
+       ex4 auto :: [auto].
+
+  .. not sure "self" and "next" do anything special.  I get the same error
+     message for both from constructs like
+
+     Ltac2 Notation "ex5" x(self) := auto.
+     ex5 match.
+
+     Syntax error: [tactic:tac2expr level 5] expected after 'match' (in [tactic:tac2expr]).
+
+  :n:`self`
+    parses an Ltac2 expression at the current level and returns it as is.
+
+  :n:`next`
+    parses an Ltac2 expression at the next level and returns it as is.
+
+  :n:`thunk(@ltac2_scope)`
+    Used for semantic effect only, parses the same as :token:`ltac2_scope`.
+    If :n:`e` is the parsed expression for :token:`ltac2_scope`, `thunk`
+    returns :n:`fun () => e`.
+
+  :n:`pattern`
+    parses a :token:`cpattern`
+
+  .. todo PR I dropped "terminal" and "keyword" because they don't seem to behave
+     any differently from just giving a string, e.g. for the keyword "match",
+     x(terminal("match")) worked the same as x("match")
```

Comment:
```suggestion
```
You added them back.  (Feel free to ignore this suggestion, though, as it's only about a comment)

Reply:
Pierre-Marie thought I should have them, I'll drop the note.